pkg/utils/addons/addons.go

```diff
@@ -27,56 +31,65 @@ func (p *pipeline) generate(pipeline unstructured.Unstructured) (unstructured.Un
-	taskDeployMap := task.([]interface{})[2].(map[string]interface{})
-	taskDeployMap["taskRef"] = map[string]interface{}{"name": "openshift-client", "kind": "ClusterTask"}
-	taskDeployMap["runAfter"] = []interface{}{"build"}
-	taskDeployMap["params"] = []interface{}{
+func kubernetesDeployTask(deployTask map[string]interface{}) map[string]interface{} {
+	deployTask["taskRef"] = map[string]interface{}{"name": "openshift-client", "kind": "ClusterTask"}
+	deployTask["runAfter"] = []interface{}{"build"}
+	deployTask["params"] = []interface{}{
 		map[string]interface{}{"name": "SCRIPT", "value": "kubectl $@"},
 		map[string]interface{}{"name": "ARGS", "value": []interface{}{"rollout", "status", "deploy/$(params.APP_NAME)"}},
 	}
-	return taskDeployMap
+	return deployTask
 }
 
-func knativeDeployTask(task interface{}) map[string]interface{} {
-	taskDeployMap := task.([]interface{})[2].(map[string]interface{})
-	taskDeployMap["name"] = "kn-service-create"
-	taskDeployMap["taskRef"] = map[string]interface{}{"name": "kn", "kind": "ClusterTask"}
-	taskDeployMap["runAfter"] = []interface{}{"build"}
+func knativeDeployTask(deployTask map[string]interface{}) map[string]interface{} {
+	deployTask["name"] = "kn-service-create"
+	deployTask["taskRef"] = map[string]interface{}{"name": "kn", "kind": "ClusterTask"}
+	deployTask["runAfter"] = []interface{}{"build"}
 
-	taskDeployMap["params"] = []interface{}{
+	deployTask["params"] = []interface{}{
 		map[string]interface{}{"name": "ARGS", "value": []interface{}{"service", "create", "$(params.APP_NAME)", "--image=$(params.IMAGE_NAME)", "--force"}},
 	}
-	return taskDeployMap
+	return deployTask
 }
 
-func CreatePipelines(template mf.Manifest, client client.Client) (mf.Manifest, error) {
-	var pipelines []unstructured.Unstructured
-
-	taskGenerators := []taskGenerator{
-		&pipeline{environment: "openshift", nameSuffix: "", generateDeployTask: openshiftDeployTask},
-		&pipeline{environment: "kubernetes", nameSuffix: "-deployment", generateDeployTask: kubernetesDeployTask},
-		&pipeline{environment: "knative", nameSuffix: "-knative", generateDeployTask: knativeDeployTask},
+func knativeResourcedDeployTask(deployTask map[string]interface{}) map[string]interface{} {
+	deployTask["name"] = "kn-service-create"
+	deployTask["taskRef"] = map[string]interface{}{"name": "kn", "kind": "ClusterTask"}
+	deployTask["runAfter"] = []interface{}{"build"}
+	deployTask["resources"] = map[string]interface{}{
+		"inputs": []interface{}{map[string]interface{}{"name": "image", "resource": "app-image", "from": []interface{}{"build"}}},
+	}
+	deployTask["params"] = []interface{}{
+		map[string]interface{}{"name": "ARGS", "value": []interface{}{"service", "create", "$(params.APP_NAME)", "--image=$(resources.inputs.image.url)", "--force"}},
 	}
+	return deployTask
+}
+
+func generateBasePipeline(template mf.Manifest, taskGenerators []taskGenerator, resourced bool) ([]unstructured.Unstructured, error) {
```

Comment:
if we decide to go on with a boolean flag `resourced bool` , i would prefer a more descriptive name.
even `usePipelineResource bool` :)